core/src/typecheck/mod.rs

```diff
@@ -2420,7 +2420,29 @@ pub fn subsumption(
     checked: UnifType,
 ) -> Result<(), UnifError> {
     let inferred_inst = instantiate_foralls(state, &mut ctxt, inferred, ForallInst::UnifVar);
-    checked.unify(inferred_inst, state, &ctxt)
+    match (inferred_inst.clone(), checked.clone()) {
+        (
+            UnifType::Concrete {
+                typ: TypeF::Record(rrows),
+                ..
+            },
+            UnifType::Concrete {
+                typ: TypeF::Dict { type_fields, .. },
+                ..
+            },
+        ) => rrows.iter().fold(
+            Ok(()),
+            |acc, row: GenericUnifRecordRowsIteratorItem<_>| -> Result<(), UnifError> {
+                match (row, &acc) {
+                    (GenericUnifRecordRowsIteratorItem::Row(a), Ok(_)) => {
+                        subsumption(state, ctxt.clone(), a.typ.clone(), *type_fields.clone())
+                    }
+                    _ => acc,
```

Comment:
Now I'm thinking about something that I haven't thought about before. During type inference, we might have open record types, such as `{foo : Number, bar : Number; tail}` where the `tail` at the end means that this value can have more fields that we don't know yet.

If we try to subtype that with a dictionary, as in `{foo : Number, bar : Number; ?a} <: {_ : Number}`, we can't let the `?a` be, because it might very much be instantiated with `baz : String` at some point. So we need to close the tail of the record.

Interestingly, we can put `Dyn` in the tail currently (instead of `?a`) to say that we don't know anything about it, that those are some fields with `Dyn` type. I wonder if we could extend the type system to add other types, for example `{foo : Number, bar : Number; Number}` (or `{foo : Number, bar : Number; _ : Number}`) to say that the tail is unknown but fields must be numbers. That would probably be a better representation.

Anyhow, I'm just thinking out loud. Let's forget that, for now, I don't think we can do better than closing the tail of the record. In practice, this means that we need to also consider the cases:

- `GenericUnifRecordRowsIteratorItem::UnifVar { id, init_level}`. In that case we need to unify this var with an empty row. Here is how it's done (**note**: it's done for _enum rows_ below, but it's exactly the same for record rows; just change any "enum" to "record"):

  https://github.com/tweag/nickel/blob/9969d9a9d40958d97f323420daba79c1b910522e/core/src/typecheck/pattern.rs#L129-L133
- `GenericUnifRecordRowsIteratorItem::TailConstant(id)`. In this case we must fail. We can raise a `WithConst` error - in the future we could have a dedicated error for better messages, but for now, that will do.